### PR TITLE
initial implementation

### PR DESCRIPTION
**Pull request by seanwiseman**
This provides the implementation for event/message publishing for the eLife bus. 

It's current use case is to publish messages via AWS SNS, though it has scope to be extended.

An example use case is shown below. This allows publishing of `ProfileMessage` via AWS SNS:

```
from bus_sdk import get_publisher

config = {
    'region': 'us-east-2',
    'subscriber': '00000000000',       
    'name': 'profile',
    'env': 'prod'
}

message = ProfileMessage(id='12345')
publisher = get_publisher(pub_name='test', pub_type='sns', config=config)
publisher.publish(message)

```

### CODE REVIEW

File: bus_sdk/__init__.py
Comment (giorgiosironi) [inline bus_sdk/__init__.py]: I think most of the users of this library will only care about passing `config`, so I would insert some sane defaults for the other two parameters, moving them later in the list for clarity

File: bus_sdk/__init__.py
Comment (giorgiosironi) [inline bus_sdk/__init__.py]: typo `uninitialzed`

File: bus_sdk/messages.py
Comment (giorgiosironi) [inline bus_sdk/messages.py]: ok, but not needed, this is produced by a PHP application

File: bus_sdk/publishers/event_publisher.py
Comment (giorgiosironi) [inline bus_sdk/publishers/event_publisher.py]: is it `dict` for a reason or should also be `Message`?

File: bus_sdk/publishers/sns_publisher.py
Comment (giorgiosironi) [inline bus_sdk/publishers/sns_publisher.py]: `overrides`

File: setup.py
Comment (giorgiosironi) [inline setup.py]: not `bussdk` due to https://www.python.org/dev/peps/pep-0008/#package-and-module-names ? Not particularly readable

File: setup.py
Comment (giorgiosironi) [inline setup.py]: link to https://github.com/elifesciences/bus somewhere wold be helpful

File: bus_sdk/__init__.py
Comment (seanwiseman) [reply bus_sdk/__init__.py]: OK I have now added defaults.

File: bus_sdk/__init__.py
Comment (seanwiseman) [reply bus_sdk/__init__.py]: 👍🏻

File: bus_sdk/messages.py
Comment (seanwiseman) [reply bus_sdk/messages.py]: Removed. `YAGNI`

File: bus_sdk/publishers/event_publisher.py
Comment (seanwiseman) [reply bus_sdk/publishers/event_publisher.py]: Was left like that initially to be more flexible, plus `Message` inherits from `dict` so syntactically it was OK. Though with the design choice of per type messages e.g. `ProfileMessage` I have now changed the abstract arg to `Message`.

File: bus_sdk/publishers/sns_publisher.py
Comment (seanwiseman) [reply bus_sdk/publishers/sns_publisher.py]: 👍🏻

File: setup.py
Comment (seanwiseman) [reply setup.py]: Yes, it was the readability that was the issue. Even though its discouraged in PEP8 underscores are still used in major package names to improve readability. An example I use a lot is http://www.django-rest-framework.org

usage:
`from rest_framework import routers, serializers`

I will always avoid '_' where possible but I think it is fine in this use case.

File: setup.py
Comment (seanwiseman) [reply setup.py]: Done.

File: bus_sdk/publishers/sns_publisher.py
Comment (giorgiosironi) [inline bus_sdk/publishers/sns_publisher.py]: small typo `AWs`

File: bus_sdk/messages.py
Comment (giorgiosironi) [inline bus_sdk/messages.py]: I think in the future we may use these message objects when reading from the bus on the other side

**Review by giorgiosironi — APPROVED** [review]

File: .gitignore
Comment (thewilkybarkid) [inline .gitignore]: Needs a new line at EOF (and has an extra line at the top)

File: bus_sdk/messages.py
Comment (thewilkybarkid) [inline bus_sdk/messages.py]: Is this needed if it's already a `dict`?

File: bus_sdk/publishers/sns_publisher.py
Comment (seanwiseman) [reply bus_sdk/publishers/sns_publisher.py]: 👍🏻

File: bus_sdk/messages.py
Comment (seanwiseman) [reply bus_sdk/messages.py]: Sounds good.

File: .gitignore
Comment (seanwiseman) [reply .gitignore]: 👍🏻

File: bus_sdk/messages.py
Comment (seanwiseman) [reply bus_sdk/messages.py]: I included it because it meant you didn't have to import `json` and pass it through `.dumps()` every time you needed the JSON representation. You can still pass it through `.dumps()` manually if preferred, but it saved me a few lines of code in modules and tests in which I didn't need `json` for anything else other than converting to the `Message` to JSON.

File: bus_sdk/publishers/sns_publisher.py
Comment (thewilkybarkid) [inline bus_sdk/publishers/sns_publisher.py]: Different implementations would return different things, so shouldn't be returning an implementation-specific response (not sure there's anything useful here?).

File: setup.py
Comment (thewilkybarkid) [inline setup.py]: Is it worth having a vendor name in there? (`elife_bus_sdk`)

File: bus_sdk/publishers/event_publisher.py
Comment (thewilkybarkid) [inline bus_sdk/publishers/event_publisher.py]: Doesn't feel needed?

File: bus_sdk/publishers/sns_publisher.py
Comment (thewilkybarkid) [reply bus_sdk/publishers/sns_publisher.py]: `EventPublisher.publish` says it's a `str` too...

File: bus_sdk/__init__.py
Comment (thewilkybarkid) [inline bus_sdk/__init__.py]: Service location? Would rather see DI...

File: bus_sdk/publishers/sns_publisher.py
Comment (seanwiseman) [reply bus_sdk/publishers/sns_publisher.py]: OK fair point, at the level of the AbstractBaseClass I will put the expected return type as `Any`.

(yes I just noticed that whilst changing this, it was the original return type, I didn't change to dict)

File: bus_sdk/publishers/event_publisher.py
Comment (thewilkybarkid) [inline bus_sdk/publishers/event_publisher.py]: It's interchanging 'message' and 'event', better to always refer to the latter?

File: setup.py
Comment (seanwiseman) [reply setup.py]: I am happy to add vendor, I was just basing it on the previous suggestion. Will add.

File: bus_sdk/publishers/event_publisher.py
Comment (seanwiseman) [reply bus_sdk/publishers/event_publisher.py]: It requires a name so that it can map the types to classes which is then used by the factory function:

```
def get_publisher_types() -> dict:
    return {pub.name: pub for pub in PUBLISHERS}
```

File: bus_sdk/publishers/event_publisher.py
Comment (seanwiseman) [reply bus_sdk/publishers/event_publisher.py]: Change `Message` to be `Event`? e.g. `ProfileEvent`...

Comment (seanwiseman) [reply bus_sdk/publishers/event_publisher.py]: `Message` is now `Event` across the package.

File: bus_sdk/__init__.py
Comment (seanwiseman) [reply bus_sdk/__init__.py]: Not sure what you mean here. 🤔

File: bus_sdk/publishers/event_publisher.py
Comment (seanwiseman) [reply bus_sdk/publishers/event_publisher.py]: @thewilkybarkid I will wait until this PR is merged before integration on Profiles, slightly fewer changes if done this way round.

File: bus_sdk/__init__.py
Comment (seanwiseman) [reply bus_sdk/__init__.py]: Have left this as is to handle not having to configure the same `publisher` in multiple modules (some other projects are more function based and this would avoid some duplication). Though you can still use this to return a new instance which can then be directly injected.

File: .pylintrc
Comment (thewilkybarkid) [inline .pylintrc]: `syntax-error` doesn't sound good!

**Review by thewilkybarkid — APPROVED** [review]
:shipit:

File: .pylintrc
Comment (seanwiseman) [reply .pylintrc]: No it is not! now removed. Was added when I was testing an alternative mock and the mock lib was throwing some strange syntax errors.

**Comment by thewilkybarkid** [comment]
I'll set up Alfred before merging.

**Pull request merged**
